### PR TITLE
GDB Plugin Upgrade part3: fs related tools

### PR DESCRIPTION
**Pull request by XuNeo**
*Note: Please adhere to [Contributing Guidelines](https://github.com/apache/nuttx/blob/master/CONTRIBUTING.md).*

## Summary

In this PR, we introduces several fs related tools. 
This PR is based on part1/2 to minimize conflicts. I'll rebase once they are merged. For now please review commits by commits.

New tools added:
1. mount: similar to device side, mount lists all the mount point.
2. `foreach_inode` print all the inode and its information
```
go through each inode and print its structure info
in the form like below:
├── i_name: bin, i_ino: 51
│   i_parent: , i_peer: data, i_child: , i_crefs: 1, i_flags: 3
│   ......(other info)
│   ├── i_name: audio, i_ino: 5
│   │   i_parent: dev, i_peer: binder, i_child: mixer......
```
3. `fdinfo` print all the file opened in each task
```
(gdb) fdinfo
PID: 0
FD  OFLAGS  POS   PATH                  BACKTRACE
0   3       0     /dev/console          0x4028ff0c <sched_backtrace+48>                 /home/neo/projects/vela2/nuttx/sched/sched/sched_backtrace.c:105
                                        0x402888f8 <file_allocate_from_tcb+236>         /home/neo/projects/vela2/nuttx/fs/inode/fs_files.c:615
                                        0x402979d0 <nx_vopen+104>                       /home/neo/projects/vela2/nuttx/fs/vfs/fs_open.c:326
                                        0x40297ad4 <nx_open+116>                        /home/neo/projects/vela2/nuttx/fs/vfs/fs_open.c:449
                                        0x40291eb4 <group_setupidlefiles+28>            /home/neo/projects/vela2/nuttx/sched/group/group_setupidlefiles.c:75
                                        0x4028df94 <nx_start+628>                       /home/neo/projects/vela2/nuttx/sched/init/nx_start.c:651
                                        0x4028119c <arm64_boot_primary_c_routine+16>    /home/neo/projects/vela2/nuttx/arch/arm64/src/common/arm64_boot.c:205

1   3       0     /dev/console          0x4028ff0c <sched_backtrace+48>                 /home/neo/projects/vela2/nuttx/sched/sched/sched_backtrace.c:105
                                        0x40289574 <file_dup3+248>                      /home/neo/projects/vela2/nuttx/fs/vfs/fs_dup2.c:177
                                        0x40288b88 <nx_dup3_from_tcb+176>               /home/neo/projects/vela2/nuttx/fs/inode/fs_files.c:314
                                        0x40288c64 <nx_dup2_from_tcb+16>                /home/neo/projects/vela2/nuttx/fs/inode/fs_files.c:901
                                        0x40288c88 <nx_dup2+28>                         /home/neo/projects/vela2/nuttx/fs/inode/fs_files.c:924
                                        0x40291ec4 <group_setupidlefiles+44>            /home/neo/projects/vela2/nuttx/sched/group/group_setupidlefiles.c:84
                                        0x4028df94 <nx_start+628>                       /home/neo/projects/vela2/nuttx/sched/init/nx_start.c:651
                                        0x4028119c <arm64_boot_primary_c_routine+16>    /home/neo/projects/vela2/nuttx/arch/arm64/src/common/arm64_boot.c:205

2   3       0     /dev/console          0x4028ff0c <sched_backtrace+48>                 /home/neo/projects/vela2/nuttx/sched/sched/sched_backtrace.c:105
                                        0x40289574 <file_dup3+248>                      /home/neo/projects/vela2/nuttx/fs/vfs/fs_dup2.c:177
                                        0x40288b88 <nx_dup3_from_tcb+176>               /home/neo/projects/vela2/nuttx/fs/inode/fs_files.c:314
                                        0x40288c64 <nx_dup2_from_tcb+16>                /home/neo/projects/vela2/nuttx/fs/inode/fs_files.c:901
                                        0x40288c88 <nx_dup2+28>                         /home/neo/projects/vela2/nuttx/fs/inode/fs_files.c:924
                                        0x40291ed0 <group_setupidlefiles+56>            /home/neo/projects/vela2/nuttx/sched/group/group_setupidlefiles.c:117
                                        0x4028df94 <nx_start+628>                       /home/neo/projects/vela2/nuttx/sched/init/nx_start.c:651
                                        0x4028119c <arm64_boot_primary_c_routine+16>    /home/neo/projects/vela2/nuttx/arch/arm64/src/common/arm64_boot.c:205
```
4. `info shm` show the share memory usage
```
(gdb) info shm
  /var/shm/xms:bq-325-1044165565 memsize: 1536000, paddr: 0x41de3970
  /var/shm/xms:bq-325-2123092606 memsize: 1536000, paddr: 0x41f5a9a8
  /var/shm/xms:fakemq-325-1835096569 memsize: 12, paddr: 0x420d19e0
(gdb)
```

Also included an performance optimization that use generate instead of list where possible.

## Impact

New feature added.

## Testing

Tested on qemu arm64(info shm is not available because it's disabled).

### CODE REVIEW

**Comment by acassis** [comment]
@XuNeo please update https://nuttx.apache.org/docs/latest/guides/gdbwithpython.html with all the steps you put here. The current "guide" is not useful and point to an external doc that will know fit to NuttX details.

File: tools/gdb/macros.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+# NOTE: GDB stores macro information based on the current stack frame's scope,
+# including the source file and line number. Therefore, there may be missing
+# macro definitions when you are at different stack frames.
+#
+# To resolve this issue, we need to retrieve all macro information from the ELF file
+# then parse and evaluate it by ourselves.
+#
+# There might be two ways to achieve this, one is to leverage the C preprocessor
+# to directly preprocess all the macros instereted into python constants
```
Comment (hartmannathan) [inline tools/gdb/macros.py:29]:
```suggestion
# to directly preprocess all the macros inserted into python constants
```

File: tools/gdb/macros.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+# gcc -E -x c -P <file_with_macros> -I/path/to/nuttx/include
+#
+# While the other way is to leverage the dwarf info stored in the ELF file,
+# with -g3 switch, we have a `.debug_macro` section containing all the information
+# about the macros.
+#
+# Currently, we using the second method.
```
Comment (hartmannathan) [inline tools/gdb/macros.py:36]:
```suggestion
# Currently, we are using the second method.
```

File: tools/gdb/macros.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        if t not in macro_map:
+            res.append(t)
+            continue
+        res += do_expand(macro_map[t], macro_map)
+
+    return res
+
+
+# NOTE: Implement a fully functional parser which can
+# preprocessing all the C marcos according to ISO 9899 standard
```
Comment (hartmannathan) [inline tools/gdb/macros.py:187]:
```suggestion
# preprocess all the C macros according to ISO 9899 standard
```

File: tools/gdb/macros.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+# may be an overkill, what we really care about are those
+# macros that can be evaluted to an constant value.
```
Comment (hartmannathan) [inline tools/gdb/macros.py:189]:
```suggestion
# macros that can be evaluated to a constant value.
```

File: tools/gdb/macros.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+#
+# However,
+# #define SUM(B,C,D) (B + C + D)
+# invoking try_expand('SUM(1,2,3)', macro_map) will give you "SUM(1,2,3)"
+#
+# We have not implemented this feature as we have not found a practical
+# use case for it in our GDB plugin.
+#
+# However, you can switch to the correct stack frame that has this macro defined
+# and let GDB expand and evaluate it for you if you really want to evalue some very
```
Comment (hartmannathan) [inline tools/gdb/macros.py:205]:
```suggestion
# and let GDB expand and evaluate it for you if you really want to evaluate some very
```

File: tools/gdb/utils.py [hunk 6]
```diff
@@ -83,6 +100,147 @@ def invoke(self, ptr, typename, elementname):
...
+def gdb_eval_or_none(expresssion):
+    """Evaluate an expression and return None if it fails"""
+    try:
+        return gdb.parse_and_eval(expresssion)
+    except gdb.error:
+        return None
+
+
+def get_symbol_value(name, locspec="nx_start", cacheable=True):
+    """Return the value of a symbol value etc: Variable, Marco"""
```
Comment (hartmannathan) [inline tools/gdb/utils.py:112]:
```suggestion
    """Return the value of a symbol value etc: Variable, Macro"""
```

File: tools/gdb/utils.py [hunk 7]
```diff
@@ -155,23 +338,37 @@ def read_ulong(buffer, offset):
 target_arch = None
 
 
-def is_target_arch(arch):
-    """Return True if the target architecture is ARCH"""
+def is_target_arch(arch, exact=False):
+    """
+    For non exactly match, this function will
```
Comment (hartmannathan) [inline tools/gdb/utils.py:343]:
```suggestion
    For non exact match, this function will
```

File: tools/gdb/macros.py (hunk 1 above)
Comment (XuNeo) [reply tools/gdb/macros.py]: Hi, these typos are fixed in PR GDB Plugin Upgrade part1. I have rebased and now they should be fixed. Thank you!

**Comment by XuNeo** [comment]
@acassis Hi Alan, 

I'm going to update the docs after uploading all my features. Is that OK? There are some command usage will be slightly changed later.

For command usage, use `help <command name>` or `<command> --help`

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**